### ISSUE TITLE
[BUG]: Inheriting from Stringable make String not Stringable

### ISSUE THREAD

**Issue by gabrieldemarmiesse** [modularml/mojo#1383]
### Bug description

Apologies if I didn't understand the feature, but from the example here: https://docs.modular.com/mojo/manual/traits.html#trait-inheritance this should work:

### Steps to reproduce

```python
trait MyStringable(Stringable):
    pass

fn to_string[T: MyStringable](a: T) -> String:
    return str(a)

def main():
    print(to_string[String](String("hello")))
```

This gives the following error: 
```
/projects/open_source/mojo-stdlib-extensions/trying_stuff.mojo:10:21: error: cannot bind type 'String' to trait 'MyStringable'
    print(to_string[String](String("hello")))
                    ^~~~~~
/projects/open_source/mojo-stdlib-extensions/trying_stuff.mojo:1:1: note: 'String' does not implement 'MyStringable'
trait MyStringable(Stringable):
^
mojo: error: failed to parse the provided Mojo
```

This is strange because this works without issue:

```mojo
alias MyStringable = Stringable

fn to_string[T: MyStringable](a: T) -> String:
    return str(a)

def main():
    print(to_string[String](String("hello")))
```

### System information

```shell
- What OS did you do install Mojo on ?
Ubuntu 22.04
$ mojo -v
mojo 0.6.0 (d55c0025)
$ modular -v
modular 0.2.0 (355ea9c0)
```

**Comment by gabrieldemarmiesse** [comment]
Many thanks for your answer. So this is because every struct must declare each trait it's implementing since the compiler can't know if it conforms to a trait or not. It's quite hard for Python devs who are used to Python's `Protocol`. But maybe it's just inexperience on my side.

I'm trying to write an equivalent of Python's list in Mojo. Since `list` has a `__str__` method that will display each element as a str, I need the list types to be

```python
trait ListElement(CollectionElement, Stringable):
    pass

@value
struct list[T: ListElement](Sized):
    ....
```

but then I can't do 

```python
var a = list[String]()
```

I'm not sure how to go about this. Is this currently impossible in mojo to make a list of printable element that can take types from the stdlib?

**Comment by gabrieldemarmiesse** [comment]
I'm very sad about that. I hope it will be possible in the future. The way traits currently work seems very rigid if we can't even implement a simple list.

**Comment by theopomies** [comment]
@soraros I raised similar questions around implementing custom traits for builtin structs, or even just structs from external modules (a la Rust) and they did confirm it will be implemented in the future https://discord.com/channels/1087530497313357884/1098713601386233997/1181287455366852758

**Comment by rarebreed** [comment]
I think one way to think about this, is that MyStringable is a subtype of Stringable, but Stringable isn't a subtype of MyStringable.  (I could be wrong about this, as covariant vs contravariance always messes me up, but so far, mojo's subtyping rules for traits appears to be covariant only).

In other words MyStingable **is a** Stringable, but the converse isn't necessarily true.  All dogs are animals, but not all animals are dogs.  Since `String` implements `Stringable`, that does not mean it implements `MyStringable`.  Currently, MyStringable hasn't added any new methods to its "contract".  But what if it did?  Whatever type you passed to `to_string` would have to be constrained by the contract specified in `MyStringable`

As for being able to add traits to existing types, ala rust traits or haskell typeclasses, they are [planning to support it](https://docs.modular.com/mojo/roadmap.html#traits-support).

```
Support for a feature like Swift’s extensions, allowing you to add a trait to a preexisting type
```

**Comment by gabrieldemarmiesse** [comment]
> @gabrieldemarmiesse What you could do for the time being, is a print that only works on `Stringable` elements:
> ```mojo
> fn print_list[T: Stringable](l: List[T]) -> String:
>   ...
> ```

I already tried this option and sadly it doesn't compile. Saying that a Stringable does no implement CollectionElement, so I think we're stuck at the moment. We must once again write the function for each type possible without code reuse (at least I didn't find a way to reuse the logic, maybe I just didn't find the right way to do it)

**Comment by lattner** [comment]
Agreed, we really only have "basic traits" so far.  There are a lot of features we need to add to build into this for it to fulfill its potential.

**Comment by affo** [comment]
Hello, I know this is closed, but this was the only way I could cope with this:

```mojo
trait StringableCollectionElement(CollectionElement, Stringable):
    ...
    # fn __moveinit__(inout self, owned existing: Self): ...
    # fn __copyinit__(inout self, existing: Self): ...
    # fn __str__(self) -> String: ...

fn append[T: CollectionElement](inout l: DynamicVector[T], e: T) -> None:
    l.append(e)

fn printv[T: StringableCollectionElement](borrowed v: DynamicVector[T]) -> None:
    let l = len(v)
    for i in range(l):
        if i == 0:
            print_no_newline("[")
        print_no_newline(v[i])
        if i < l - 1:
            print_no_newline(", ")
        else:
            print("]")

@value
struct MyInt(StringableCollectionElement):
    var wrapped: Int

    fn __str__(self) -> String:
        return str(self.wrapped)

    fn value(self) -> Int:
        return self.wrapped

fn main():
    var li = DynamicVector[MyInt]()
    append(li, MyInt(1))
    append(li, MyInt(2))
    append(li, MyInt(3))
    printv[MyInt](li)
    append(li, MyInt(5))
    printv[MyInt](li)

```

**Comment by laszlokindrat** [comment]
@affo Please open a new ticket with this request, so we can triage and discuss with the team! :)

**Comment by affo** [comment]
@laszlokindrat thank you!

Ended up creating a discussion as I could not find this in the roadmap: https://github.com/modularml/mojo/discussions/1681.
